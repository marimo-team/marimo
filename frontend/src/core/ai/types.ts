/* Copyright 2024 Marimo. All rights reserved. */

export interface ChatAttachment {
  /**
   * The name of the attachment, usually the file name.
   */
  name?: string;
  /**
   * A string indicating the [media type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type).
   * By default, it's extracted from the pathname's extension.
   */
  contentType?: string;
  /**
   * The URL of the attachment. It can either be a URL to a hosted file or a [Data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs).
   */
  url: string;
}
