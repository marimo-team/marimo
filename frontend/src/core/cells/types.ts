/* Copyright 2024 Marimo. All rights reserved. */
import { OutputMessage } from "../kernel/messages";
import { SerializedEditorState } from "../codemirror/types";
import { Outline } from "./outline";
import { CellId } from "./ids";
import { DEFAULT_CELL_NAME } from "./names";
import { Milliseconds, Seconds } from "@/utils/time";

/**
 * The status of a cell.
 *
 * queued: queued by the kernel.
 * running: currently executing.
 * idle: not running.
 * stale: stale, likely because the cell was disabled or an ancestor was disabled.
 * disabled-transitively: disabled because an ancestor was disabled.
 */
export type CellStatus =
  | "queued"
  | "running"
  | "idle"
  | "stale"
  | "disabled-transitively";

/**
 * Create a new cell with default state.
 */
export function createCell({
  id,
  name = DEFAULT_CELL_NAME,
  code = "",
  lastCodeRun = null,
  edited = false,
  config = {},
  serializedEditorState = null,
}: Partial<CellData> & { id: CellId }): CellData {
  return {
    id: id,
    config: config,
    name: name,
    code: code,
    edited: edited,
    lastCodeRun: lastCodeRun,
    serializedEditorState: serializedEditorState,
  };
}

export function createCellRuntimeState(): CellRuntimeState {
  return {
    outline: null,
    output: null,
    consoleOutputs: [],
    status: "idle",
    interrupted: false,
    errored: false,
    stopped: false,
    runElapsedTimeMs: null,
    runStartTimestamp: null,
    debuggerActive: false,
  };
}

/**
 * Data of the cell
 */
export interface CellData {
  id: CellId;
  /** user-given name, or default */
  name: string;
  /** current contents of the editor */
  code: string;
  /** whether this cell has been modified since its last run */
  edited: boolean;
  /** snapshot of code that was last run */
  lastCodeRun: string | null;
  /** cell configuration */
  config: CellConfig;
  /** serialized state of the underlying editor */
  serializedEditorState: SerializedEditorState | null;
}

export interface CellRuntimeState {
  /** a message encoding the cell's output */
  output: OutputMessage | null;
  /** TOC outline */
  outline: Outline | null;
  /** messages encoding the cell's console outputs. */
  consoleOutputs: OutputMessage[];
  /** current status of the cell */
  status: CellStatus;
  /** whether this cell has been interrupted since its last run */
  interrupted: boolean;
  /** whether this cell was stopped with mo.stop */
  stopped: boolean;
  /**
   * whether marimo encountered an error when trying to register or run
   * this cell (such as a multiple definition error)
   */
  errored: boolean;
  /** run start time, as seconds since epoch */
  runStartTimestamp: Seconds | null;
  /** run elapsed time, in milliseconds */
  runElapsedTimeMs: Milliseconds | null;
  /** debugger active */
  debuggerActive: boolean;
}

export interface CellConfig {
  /**
   * If true, the cell and its descendants are unable to run.
   */
  disabled?: boolean;
  /**
   * If true, the cell's code is hidden from the notebook.
   */
  hide_code?: boolean;
}
