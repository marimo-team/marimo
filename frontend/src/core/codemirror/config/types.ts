/* Copyright 2024 Marimo. All rights reserved. */

/**
 * The type of placeholder to use for the editor.
 *
 * - "marimo-import" -> Shows `import marimo as mo`
 * - "ai" -> Shows `Start coding or Generate with AI`
 * - "none" -> No placeholder
 */
export type PlaceholderType = "marimo-import" | "ai" | "none";
