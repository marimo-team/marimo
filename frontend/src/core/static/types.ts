/* Copyright 2024 Marimo. All rights reserved. */
import { Base64String } from "@/utils/json/base64";
import { NotebookState } from "../cells/cells";
import { CellId } from "../cells/ids";

export interface StaticNotebookState extends Pick<NotebookState, "cellIds"> {
  version: string;
  assetUrl: string;
  cellRuntime: Record<CellId, Base64String>;
  cellData: Record<CellId, Base64String>;
}

export type StaticVirtualFiles = Record<string, { base64: Base64String }>;

export interface MarimoStaticState {
  version: string;
  notebookState: StaticNotebookState;
  assetUrl: string;
  files: StaticVirtualFiles;
}
