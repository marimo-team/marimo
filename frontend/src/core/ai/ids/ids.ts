/* Copyright 2024 Marimo. All rights reserved. */

import type { TypedString } from "@/utils/typed";

export const PROVIDERS = [
  "openai",
  "anthropic",
  "google",
  "ollama",
  "bedrock",
  "deepseek",
  "azure",
] as const;
export type ProviderId = (typeof PROVIDERS)[number];

export type ShortModelId = TypedString<"ShortModelId">;

/**
 * Qualified model id
 * `provider_id/short_model_id`;
 */
export type QualifiedModelId = `${ProviderId}/${ShortModelId}`;

export class AiModelId {
  readonly providerId: ProviderId;
  readonly shortModelId: ShortModelId;

  constructor(providerId: ProviderId, shortModelId: ShortModelId) {
    this.providerId = providerId;
    this.shortModelId = shortModelId;
  }

  get id(): QualifiedModelId {
    return `${this.providerId}/${this.shortModelId}`;
  }

  static parse(id: string): AiModelId {
    if (!id.includes("/")) {
      const providerId = guessProviderId(id);
      return new AiModelId(providerId, id as ShortModelId);
    }

    const [providerId, ...shortModelId] = id.split("/");
    return new AiModelId(
      providerId as ProviderId,
      shortModelId.join("/") as ShortModelId,
    );
  }
}

function guessProviderId(id: string): ProviderId {
  if (id.startsWith("gpt") || id.startsWith("o3") || id.startsWith("o1")) {
    return "openai";
  }
  if (id.startsWith("claude")) {
    return "anthropic";
  }
  if (id.startsWith("gemini") || id.startsWith("google")) {
    return "google";
  }
  if (id.startsWith("deepseek")) {
    return "deepseek";
  }
  return "ollama";
}

export function isKnownAIProvider(providerId: ProviderId): boolean {
  return PROVIDERS.includes(providerId);
}
