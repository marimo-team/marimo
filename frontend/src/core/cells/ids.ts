/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-redeclare */

import { invariant } from "@/utils/invariant";
import { TypedString } from "../../utils/typed";

const lowercase = "abcdefghijklmnopqrstuvwxyz";
const uppercase = lowercase.toUpperCase();
const alphabet = lowercase + uppercase;

/**
 * A typed CellId
 */
export type CellId = TypedString<"CellId">;
export const CellId = {
  /**
   * Create a new CellId, a random 4 letter string.
   */
  create(): CellId {
    let id = "";
    for (let i = 0; i < 4; i++) {
      id += alphabet[Math.floor(Math.random() * alphabet.length)];
    }
    return id as CellId;
  },
};

/**
 * The id of an HTML element that represents a cell
 */
export type HTMLCellId = `cell-${CellId}`;
export const HTMLCellId = {
  /**
   * Create a new HTMLCellId
   */
  create(cellId: CellId): HTMLCellId {
    invariant(cellId != null, "cellId is required");

    return `cell-${cellId}`;
  },
  parse(htmlCellId: HTMLCellId): CellId {
    return htmlCellId.slice(5) as CellId;
  },
  /**
   * Get the cell container ancestor of `element`, if any.
   */
  findElement(element: Element): (Element & { id: HTMLCellId }) | null {
    return element.closest('div[id^="cell-"]');
  },
};

/**
 * A typed UIElementId
 */
export type UIElementId = `${CellId}-${string}`;
export const UIElementId = {
  parse(element: Element): UIElementId | null {
    return element.getAttribute("object-id") as UIElementId | null;
  },
  parseOrThrow(element: Element): UIElementId {
    const id = UIElementId.parse(element);
    invariant(id, "<marimo-ui-element/> missing object-id attribute");
    return id;
  },
};
