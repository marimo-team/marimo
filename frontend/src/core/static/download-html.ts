/* Copyright 2024 Marimo. All rights reserved. */
import { Objects } from "@/utils/objects";
import { NotebookState, getNotebook } from "../cells/cells";
import { getMarimoVersion } from "../dom/marimo-tag";
import { downloadVirtualFiles } from "./files";
import { StaticVirtualFiles } from "./types";
import { serializeJsonToBase64 } from "@/utils/json/base64";
import { readCode } from "../network/requests";
import { downloadBlob } from "@/utils/download";
import { Paths } from "@/utils/paths";

// For Testing:
// Flip this to `true` to use local assets instead of CDN
// This should be `false` for production
const ENABLE_LOCAL_ASSETS = false;

/**
 * Downloads the current notebook as an HTML file.
 */
export async function createStaticHTMLNotebook() {
  const notebook = getNotebook();
  const version = getMarimoVersion();

  const assetUrl =
    ENABLE_LOCAL_ASSETS ||
    process.env.NODE_ENV === "development" ||
    process.env.NODE_ENV === "test"
      ? window.location.origin
      : `https://cdn.jsdelivr.net/npm/@marimo-team/frontend@${version}/dist`;

  const codeResponse = await readCode();

  const html = constructHTML({
    notebookState: notebook,
    version: version,
    assetUrl: assetUrl,
    existingDocument: document,
    files: await downloadVirtualFiles(),
    code: codeResponse.contents,
  });

  return html;
}

/**
 * Downloads the current notebook as an HTML file.
 */
export async function downloadAsHTML(opts: { filename: string }) {
  const { filename } = opts;
  const html = await createStaticHTMLNotebook();

  const filenameWithoutPath = Paths.basename(filename) ?? "notebook.py";
  const filenameWithoutExtension =
    filenameWithoutPath.split(".").shift() ?? "app";

  downloadBlob(
    new Blob([html], { type: "text/html" }),
    `${filenameWithoutExtension}.html`,
  );
}

/**
 * Constructs HTML for the a static rendering of the page.
 */
export function constructHTML(opts: {
  version: string;
  notebookState: Pick<NotebookState, "cellIds" | "cellData" | "cellRuntime">;
  assetUrl: string;
  files: StaticVirtualFiles;
  existingDocument: Document;
  code: string;
}) {
  const { version, notebookState, assetUrl, existingDocument, files, code } =
    opts;

  const staticHead = existingDocument.head.cloneNode(true) as HTMLHeadElement;
  // Remove fonts, that contain as="font"
  staticHead.querySelectorAll("link[as=font]").forEach((el) => el.remove());

  const elementsFromBody = [
    "marimo-filename",
    "marimo-version",
    "marimo-user-config",
    "marimo-app-config",
  ];

  // Create initial elements
  const html = `
  <!DOCTYPE html>
  <html>
    <head>
      <base href="/">
      ${staticHead.innerHTML}
    </head>

    <body>
      <div id="root"></div>
      <marimo-mode data-mode="read" hidden=""></marimo-mode>
      ${elementsFromBody
        .map((id) => {
          const element = existingDocument.querySelector(id);
          if (!element) {
            throw new Error(`Element ${id} not found.`);
          }
          return element.outerHTML;
        })
        .join("\n")}
      <script data-marimo="true">
        window.__MARIMO_STATIC__ = {};
        window.__MARIMO_STATIC__.version = "${version}";
        window.__MARIMO_STATIC__.notebookState = ${JSON.stringify({
          cellIds: notebookState.cellIds,
          cellData: Objects.mapValues(
            notebookState.cellData,
            serializeJsonToBase64,
          ),
          cellRuntime: Objects.mapValues(
            notebookState.cellRuntime,
            serializeJsonToBase64,
          ),
        })};
        window.__MARIMO_STATIC__.assetUrl = "${assetUrl}";
        window.__MARIMO_STATIC__.files = ${JSON.stringify(files)};
      </script>

      <marimo-code hidden="">
        ${encodeURIComponent(code)}
      </marimo-code>
    </body>
  </html>
  `;

  const doc = new DOMParser().parseFromString(html, "text/html");

  // Copy over any stylesheets or scripts from the body that are URL based
  const body = existingDocument.body;
  const appendToBody = (element: Element) => {
    const el = element.cloneNode(true);
    if (el instanceof HTMLElement) {
      el.setAttribute("crossorigin", "anonymous");
    }
    doc.body.append(element.cloneNode(true));
  };
  body.querySelectorAll("style[href]").forEach(appendToBody);
  body.querySelectorAll("script[src]").forEach(appendToBody);
  body.querySelectorAll("link[href]").forEach(appendToBody);

  // Update scripts to point to CDN
  const scripts = doc.querySelectorAll("script");
  scripts.forEach((script) => {
    const src = script.getAttribute("src");
    if (src) {
      script.setAttribute("src", updateAssetUrl(src, assetUrl));
      script.setAttribute("crossorigin", "anonymous");
    } else if (!Object.hasOwn(script.dataset, "marimo")) {
      script.remove();
    }
  });

  // Update links to point to CDN
  const links = doc.querySelectorAll("link");
  links.forEach((link) => {
    const href = link.getAttribute("href");
    if (href) {
      link.setAttribute("href", updateAssetUrl(href, assetUrl));
      link.setAttribute("crossorigin", "anonymous");
    } else {
      link.remove();
    }
  });

  // Remove style tags
  const styles = doc.querySelectorAll("style");
  styles.forEach((style) => style.remove());

  // Handle testing with Vite
  if (process.env.NODE_ENV === "development") {
    const devScript = doc.createElement("script");
    devScript.setAttribute("type", "module");
    devScript.setAttribute("src", `${window.location.origin}/src/main.tsx`);
    doc.body.append(devScript);
  }

  doc.head.innerHTML = `
  ${doc.head.innerHTML}
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Fira+Mono:wght@400;500;700&family=Lora&family=PT+Sans:wght@400;700&display=swap" rel="stylesheet">
  `.trim();

  // add DOCTYPE back in
  return `<!DOCTYPE html>\n${doc.documentElement.outerHTML}`;
}

function updateAssetUrl(existingUrl: string, assetBaseUrl: string) {
  // Will convert: https://localhost:8080/assets/index-c78b8d10.js
  //  Or will convert ./assets/index-c78b8d10.js
  //  Or will convert /assets/index-c78b8d10.js
  // into: https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.43/dist/assets/index-c78b8d10.js

  // relative './...'
  if (existingUrl.startsWith("./")) {
    return `${assetBaseUrl}${existingUrl.slice(1)}`;
  }

  // relative '/...'
  if (existingUrl.startsWith("/")) {
    return `${assetBaseUrl}${existingUrl}`;
  }

  // absolute path
  const url = new URL(existingUrl);
  if (url.origin !== window.location.origin) {
    return `${assetBaseUrl}${url.pathname}`;
  }

  // otherwise, leave as is
  return existingUrl;
}

export const visibleForTesting = {
  updateAssetUrl,
};
