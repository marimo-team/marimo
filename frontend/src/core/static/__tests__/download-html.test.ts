/* Copyright 2024 Marimo. All rights reserved. */
import { describe, expect, it } from "vitest";
import { constructHTML, visibleForTesting } from "../download-html";
import { CellId } from "@/core/cells/ids";
import { createCell, createCellRuntimeState } from "@/core/cells/types";
import { JSDOM } from "jsdom";
import prettier from "prettier";
import { Base64String } from "@/utils/json/base64";

const { updateAssetUrl } = visibleForTesting;

const DOC = `
<html lang="en">
  <head>
    <style id="random-style">
      #thing {
        visibility: hidden;
      }
    </style>
    <meta charset="utf-8" />
    <link rel="icon" href="/favicon.ico" />
    <link rel="preload" href="/assets/gradient-6c6e9bb3.png" as="image" />
    <link rel="preload" href="/assets/noise-b5c8172e.png" as="image" />
    <link
      rel="preload"
      href="/assets/Lora-VariableFont_wght-dqKLWsPt.ttf"
      as="font"
      crossorigin="anonymous"
    />
    <link
      rel="preload"
      href="/assets/PTSans-Regular-sS9EvFu5.ttf"
      as="font"
      crossorigin="anonymous"
    />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="theme-color" content="#000000" />
    <meta name="description" content="a marimo app" />
    <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
    <link rel="manifest" href="/manifest.json" />
    <script data-marimo="true">
      function __resizeIframe(obj) {
        // Resize the iframe to the height of the content
        obj.style.height =
          obj.contentWindow.document.documentElement.scrollHeight + "px";

        // Resize the iframe when the content changes
        const resizeObserver = new ResizeObserver((entries) => {
          obj.style.height =
            obj.contentWindow.document.documentElement.scrollHeight + "px";
        });
        resizeObserver.observe(obj.contentWindow.document.body);
      }
    </script>
  </head>
  <body class="light light-theme">
    <marimo-filename hidden="">app.py</marimo-filename>
    <marimo-mode data-mode="edit" hidden=""></marimo-mode>
    <marimo-version data-version="0.1.44" hidden=""></marimo-version>
    <marimo-server-token data-token="123" hidden=""></marimo-server-token>
    <marimo-user-config
      data-config='{"display": {"theme": "light"}, "keymap": {"preset": "default"}, "save": {"autosave": "after_delay", "format_on_save": true, "autosave_delay": 1000}, "server": {"browser": "default"}, "experimental": {"layouts": true, "theming": true}, "runtime": {"auto_instantiate": true}, "completion": {"activate_on_typing": true, "copilot": true}}'
      hidden=""
    ></marimo-user-config>
    <marimo-app-config data-config='{"width": "full", "layout_file": null}' hidden=""></marimo-app-config>

    <title>app</title>
    <script type="module" crossorigin="" src="/assets/index-2e140915.js"></script>
    <link rel="stylesheet" href="/assets/index-116d8088.css" />

    <div id="root">
      <main>App HTML should not render</main>
    </div>
  </body>
</html>
`;

describe("download-html", () => {
  it("should construct html correctly", async () => {
    const cellId = "1" as CellId;
    const version = "0.1.44";
    const result = constructHTML({
      version: version,
      notebookState: {
        cellIds: [cellId],
        cellData: {
          [cellId]: createCell({
            id: cellId,
            name: "cell",
            code: "code",
            config: {
              hide_code: true,
            },
          }),
        },
        cellRuntime: { [cellId]: createCellRuntimeState() },
      },
      files: {
        "/@file/note.txt": {
          base64: "data:text/plain;base64,bm90ZQo=" as Base64String,
        },
      },
      code: "import marimo as mo\n\nmo.html('<h1>hello</h1>')",
      assetUrl: `https://cdn.jsdelivr.net/npm/@marimo-team/frontend@${version}/dist`,
      existingDocument: new JSDOM(DOC).window.document,
    });

    const formattedResult = await prettier.format(result, {
      parser: "html",
    });
    expect(formattedResult).toMatchSnapshot();
  });
});

describe("updateAssetUrl", () => {
  const assetBaseUrl =
    "https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.43/dist";

  it('should convert relative URL starting with "./"', () => {
    const existingUrl = "./assets/index-c78b8d10.js";
    const expected = `${assetBaseUrl}/assets/index-c78b8d10.js`;
    expect(updateAssetUrl(existingUrl, assetBaseUrl)).toBe(expected);
  });

  it('should convert relative URL starting with "/"', () => {
    const existingUrl = "/assets/index-c78b8d10.js";
    const expected = `${assetBaseUrl}/assets/index-c78b8d10.js`;
    expect(updateAssetUrl(existingUrl, assetBaseUrl)).toBe(expected);
  });

  it("should convert absolute URL from a different origin", () => {
    const existingUrl = "https://localhost:8080/assets/index-c78b8d10.js";
    const expected = `${assetBaseUrl}/assets/index-c78b8d10.js`;
    expect(updateAssetUrl(existingUrl, assetBaseUrl)).toBe(expected);
  });

  it("should not modify URL from the same origin", () => {
    // Assuming window.location.origin is 'https://localhost:8080'
    const existingUrl = "https://localhost:8080/assets/index-c78b8d10.js";
    // Mock window.location.origin to match the existingUrl's origin
    Object.defineProperty(window, "location", {
      value: {
        origin: "https://localhost:8080",
      },
    });

    expect(updateAssetUrl(existingUrl, assetBaseUrl)).toBe(existingUrl);
  });
});
