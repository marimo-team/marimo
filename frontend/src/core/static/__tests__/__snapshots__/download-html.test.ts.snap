// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`download-html > should construct html correctly 1`] = `
"<!doctype html>
<html>
  <head>
    <base href="/" />

    <meta charset="utf-8" />
    <link
      rel="icon"
      href="https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.44/dist/favicon.ico"
      crossorigin="anonymous"
    />
    <link
      rel="preload"
      href="https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.44/dist/assets/gradient-6c6e9bb3.png"
      as="image"
      crossorigin="anonymous"
    />
    <link
      rel="preload"
      href="https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.44/dist/assets/noise-b5c8172e.png"
      as="image"
      crossorigin="anonymous"
    />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="theme-color" content="#000000" />
    <meta name="description" content="a marimo app" />
    <link
      rel="apple-touch-icon"
      href="https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.44/dist/apple-touch-icon.png"
      crossorigin="anonymous"
    />
    <link
      rel="manifest"
      href="https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.44/dist/manifest.json"
      crossorigin="anonymous"
    />
    <script data-marimo="true">
      function __resizeIframe(obj) {
        // Resize the iframe to the height of the content
        obj.style.height =
          obj.contentWindow.document.documentElement.scrollHeight + "px";

        // Resize the iframe when the content changes
        const resizeObserver = new ResizeObserver((entries) => {
          obj.style.height =
            obj.contentWindow.document.documentElement.scrollHeight + "px";
        });
        resizeObserver.observe(obj.contentWindow.document.body);
      }
    </script>

    <link rel="preconnect" href="https://fonts.googleapis.com" />
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="" />
    <link
      href="https://fonts.googleapis.com/css2?family=Fira+Mono:wght@400;500;700&amp;family=Lora&amp;family=PT+Sans:wght@400;700&amp;display=swap"
      rel="stylesheet"
    />
  </head>

  <body>
    <div id="root"></div>
    <marimo-mode data-mode="read" hidden=""></marimo-mode>
    <marimo-filename hidden="">app.py</marimo-filename>
    <marimo-version data-version="0.1.44" hidden=""></marimo-version>
    <marimo-user-config
      data-config='{"display": {"theme": "light"}, "keymap": {"preset": "default"}, "save": {"autosave": "after_delay", "format_on_save": true, "autosave_delay": 1000}, "server": {"browser": "default"}, "experimental": {"layouts": true, "theming": true}, "runtime": {"auto_instantiate": true}, "completion": {"activate_on_typing": true, "copilot": true}}'
      hidden=""
    ></marimo-user-config>
    <marimo-app-config
      data-config='{"width": "full", "layout_file": null}'
      hidden=""
    ></marimo-app-config>
    <script data-marimo="true">
      window.__MARIMO_STATIC__ = {};
      window.__MARIMO_STATIC__.version = "0.1.44";
      window.__MARIMO_STATIC__.notebookState = {
        cellIds: ["1"],
        cellData: {
          1: "JTdCJTIyaWQlMjIlM0ElMjIxJTIyJTJDJTIyY29uZmlnJTIyJTNBJTdCJTIyaGlkZV9jb2RlJTIyJTNBdHJ1ZSU3RCUyQyUyMm5hbWUlMjIlM0ElMjJjZWxsJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvZGUlMjIlMkMlMjJlZGl0ZWQlMjIlM0FmYWxzZSUyQyUyMmxhc3RDb2RlUnVuJTIyJTNBbnVsbCUyQyUyMnNlcmlhbGl6ZWRFZGl0b3JTdGF0ZSUyMiUzQW51bGwlN0Q=",
        },
        cellRuntime: {
          1: "JTdCJTIyb3V0bGluZSUyMiUzQW51bGwlMkMlMjJvdXRwdXQlMjIlM0FudWxsJTJDJTIyY29uc29sZU91dHB1dHMlMjIlM0ElNUIlNUQlMkMlMjJzdGF0dXMlMjIlM0ElMjJpZGxlJTIyJTJDJTIyaW50ZXJydXB0ZWQlMjIlM0FmYWxzZSUyQyUyMmVycm9yZWQlMjIlM0FmYWxzZSUyQyUyMnN0b3BwZWQlMjIlM0FmYWxzZSUyQyUyMnJ1bkVsYXBzZWRUaW1lTXMlMjIlM0FudWxsJTJDJTIycnVuU3RhcnRUaW1lc3RhbXAlMjIlM0FudWxsJTJDJTIyZGVidWdnZXJBY3RpdmUlMjIlM0FmYWxzZSU3RA==",
        },
      };
      window.__MARIMO_STATIC__.assetUrl =
        "https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.44/dist";
      window.__MARIMO_STATIC__.files = {
        "/@file/note.txt": { base64: "data:text/plain;base64,bm90ZQo=" },
      };
    </script>

    <marimo-code hidden="">
      import%20marimo%20as%20mo%0A%0Amo.html('%3Ch1%3Ehello%3C%2Fh1%3E')
    </marimo-code>

    <script
      type="module"
      crossorigin="anonymous"
      src="https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.44/dist/assets/index-2e140915.js"
    ></script>
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.44/dist/assets/index-116d8088.css"
      crossorigin="anonymous"
    />
  </body>
</html>
"
`;
